// @wpeform/react/dist/sdk/form.js
import { jsx } from 'react/jsx-runtime';
import { lazy } from 'react';
import { RenderSsr } from './render-ssr.js';
export { p as parseStyleData, s as setupWPEFormGlobals } from '../setup-chunk-wpeform-react-9294b13c.js';
export { B as ButtonIconPositionEnum, a as ButtonOpenTypeEnum, C as ConsequenceActionEnum, D as DarkThemeModeEnum, E as EventComparisonEnum, b as EventHasEnum, c as EventOperationEnum, d as EventRelationEnum, F as FormElementCategoryEnum, e as FormPaginationRestrictionEnum, f as FormPaymentTypeEnum, g as FormPaymentsCouponTypeEnum, h as FormSettingsBackgroundAttachmentEnum, i as FormSettingsBackgroundOriginAndClipEnum, j as FormSettingsBackgroundRepeatEnum, k as FormSettingsLimitOnLoginEnum, G as GenericAlignmentEnum, l as GenericSizeEnum, H as HeadingSizeEnum, m as HeadingTagEnum, M as MaskTypeEnum, N as NumberComparisonEnum, O as OptionColumnEnum, P as PaginationTypeEnum, n as PrefilEnum, R as RedirectTypeEnum, o as ResourceViewModeEnum, S as ScoreFromToTypeEnum, p as ScoreOperationEnum, q as SelectTypeEnum, r as SettingsAppearanceContainerLayoutEnum, s as SettingsAppearanceControlAlignmentEnum, t as SettingsAppearanceControlLayoutEnum, u as SettingsAppearanceControlTypeEnum, v as SettingsAppearanceProgressBarPositionEnum, w as SubmissionStatIntervalEnum, x as SubmitTimerEnum, T as TrashEnum, V as ValidationFilterTypeEnum, W as WidthPresetsEnum } from '../index-chunk-wpeform-react-daf2eca0.js';
import '../index-chunk-wpeform-react-b4d17b34.js';
import 'styled-components';
import 'react-dom/server';
import 'react-is';
import 'query-string';
import '@react-spring/web';
import 'react-intersection-observer';
import 'zustand';
import '@fortawesome/fontawesome-svg-core';
import '@fortawesome/react-fontawesome';
import 'react-dom';

const LazyForm = lazy(() => import('./lazy-imports.js'));
/**
 * WPEForm component, for rendering a WPEForm. This assumes the globals are
 * available in the window.WPEFormGraphQL App. See the documentation for more
 * info.
 *
 * @param props Component props.
 */
function WPEForm(props) {
    const { formId, controls, panels, themeStyle, inShadow = false, overrideDarkMode, } = props;
    return (jsx(RenderSsr, Object.assign({ controls: controls, inShadow: inShadow, panels: panels, themeStyle: themeStyle }, { children: jsx(LazyForm, { id: formId.toString(), controls: controls, inShadow: inShadow, mode: "newsubmission", panels: panels, ssr: false, themeStyle: themeStyle, overrideDarkMode: overrideDarkMode }, void 0) }), void 0));
}

export { WPEForm };
