// @wpeform/react/dist/sdk/noLazy.js
import { jsx } from 'react/jsx-runtime';
import { ApolloProvider } from '@apollo/client';
import { a as apolloClient, F as Form } from '../lazy-imports-chunk-wpeform-react-d9d32ae1.js';
import { y as Shadow, z as FormSkeleton } from '../index-chunk-wpeform-react-daf2eca0.js';
export { B as ButtonIconPositionEnum, a as ButtonOpenTypeEnum, C as ConsequenceActionEnum, D as DarkThemeModeEnum, E as EventComparisonEnum, b as EventHasEnum, c as EventOperationEnum, d as EventRelationEnum, F as FormElementCategoryEnum, e as FormPaginationRestrictionEnum, f as FormPaymentTypeEnum, g as FormPaymentsCouponTypeEnum, h as FormSettingsBackgroundAttachmentEnum, i as FormSettingsBackgroundOriginAndClipEnum, j as FormSettingsBackgroundRepeatEnum, k as FormSettingsLimitOnLoginEnum, G as GenericAlignmentEnum, l as GenericSizeEnum, H as HeadingSizeEnum, m as HeadingTagEnum, M as MaskTypeEnum, N as NumberComparisonEnum, O as OptionColumnEnum, P as PaginationTypeEnum, n as PrefilEnum, R as RedirectTypeEnum, o as ResourceViewModeEnum, S as ScoreFromToTypeEnum, p as ScoreOperationEnum, q as SelectTypeEnum, r as SettingsAppearanceContainerLayoutEnum, s as SettingsAppearanceControlAlignmentEnum, t as SettingsAppearanceControlLayoutEnum, u as SettingsAppearanceControlTypeEnum, v as SettingsAppearanceProgressBarPositionEnum, w as SubmissionStatIntervalEnum, x as SubmitTimerEnum, T as TrashEnum, V as ValidationFilterTypeEnum, W as WidthPresetsEnum } from '../index-chunk-wpeform-react-daf2eca0.js';
import { F as FallbackContainer } from '../index-chunk-wpeform-react-b4d17b34.js';
export { p as parseStyleData, s as setupWPEFormGlobals } from '../setup-chunk-wpeform-react-9294b13c.js';
import '@fortawesome/fontawesome-svg-core';
import '@fortawesome/free-brands-svg-icons';
import '@fortawesome/free-solid-svg-icons';
import '@fortawesome/free-regular-svg-icons';
import 'react';
import 'query-string';
import 'styled-components';
import '../index-chunk-wpeform-react-081736f4.js';
import '@react-spring/web';
import '@use-gesture/react';
import 'react-dom';
import 'react-dom/server';
import 'react-is';
import 'react-intersection-observer';
import 'zustand';
import '@fortawesome/react-fontawesome';

/**
 * WPEForm component, for rendering a WPEForm. This assumes the globals are
 * available in the window.WPEFormGraphQL App. See the documentation for more
 * info.
 *
 * @param props Component props.
 */
function WPEForm(props) {
    const { formId, controls, panels, themeStyle, inShadow = false } = props;
    const isSSR = typeof window === 'undefined';
    const loadingUi = (jsx(Shadow, Object.assign({ inShadow: inShadow, ssr: true }, { children: jsx(FallbackContainer, Object.assign({ themeStyle: themeStyle, ssr: true }, { children: jsx(FormSkeleton, { panels: panels, controls: controls }, void 0) }), void 0) }), void 0));
    // check for SSR
    if (isSSR) {
        return loadingUi;
    }
    return (jsx(ApolloProvider, Object.assign({ client: apolloClient }, { children: jsx(Form, { id: formId.toString(), controls: controls, inShadow: inShadow, mode: "newsubmission", panels: panels, ssr: false, themeStyle: themeStyle }, void 0) }), void 0));
}

export { WPEForm };
